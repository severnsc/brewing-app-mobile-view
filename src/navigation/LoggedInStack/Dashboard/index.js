import React from "react";
import { Title } from "../../../components";

const dashboard = {
  DASHBOARD: {
    screen: () => null,
    navigationOptions: {
      title: "Home",
      headerTitle: ({ allowFontScaling, style, children }) => (
        <Title
          value={children}
          style={style}
          allowFontScaling={allowFontScaling}
          testID="createAccountTitle"
          color="white"
        />
      )
    }
  }
};

export default dashboard;
