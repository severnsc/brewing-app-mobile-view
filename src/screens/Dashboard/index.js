import React, { useState } from "react";
import { FlatList, View } from "react-native";
import { Card, Text, Tabs, Tab, Subtitle } from "../../components";
import { currency } from "../../helpers";

const Dashboard = ({ needsToBeReordered, upcomingDeliveries }) => {
  const total = needsToBeReordered.reduce(
    (acc, cv) => acc + cv.unitCost * cv.reorderQuantity,
    0
  );
  const [active, setActive] = useState(true);
  const renderItem = active => {
    if (active) {
      return item => (
        <Card
          upperLeft={<Subtitle value={item.object.name} />}
          upperRight={<Text value={currency(item.costUnit) + item.unitCost} />}
          lowerLeft={
            <Text
              value={
                "Remaining: " +
                item.currentQuantity +
                (item.quantityUnit ? item.quantityUnit : "")
              }
            />
          }
          lowerRight={
            <Text
              value={
                "Reorder amount: " +
                item.reorderQuantity +
                (item.quantityUnit ? item.quantityUnit : "")
              }
            />
          }
        />
      );
    } else {
      return item => (
        <Card
          upperLeft={item.object.name}
          upperRight={item.deliveryDate}
          lowerLeft={
            "Amount ordered: " +
            item.reorderQuantity +
            (item.quantityUnit ? item.quantityUnit : "")
          }
          lowerRight={
            "Remaining: " +
            item.currentQuantity +
            (item.quantityUnit ? item.quantityUnit : "")
          }
        />
      );
    }
  };
  const listFooter = active ? (
    <Text>
      Total reorder cost: {currency(needsToBeReordered[0].costUnit) + total}
    </Text>
  ) : null;
  return (
    <View>
      <Tabs>
        {() => {
          return (
            <React.Fragment>
              <Tab
                value="Needs to be reordered"
                active={active}
                onPress={() => setActive(!active)}
              />
              <Tab
                value="Upcoming deliveries"
                active={!active}
                onPress={() => setActive(!active)}
              />
            </React.Fragment>
          );
        }}
      </Tabs>
      <FlatList
        data={active ? needsToBeReordered : upcomingDeliveries}
        renderItem={renderItem(active)}
        keyExtractor={item => item.id}
        ListFooterComponent={listFooter}
      />
    </View>
  );
};

Dashboard.defaultProps = {
  needsToBeReordered: [{ costUnit: "" }],
  upcomingDeliveries: []
};

export default Dashboard;
