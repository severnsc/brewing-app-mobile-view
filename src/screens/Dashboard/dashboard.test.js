import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Dashboard from ".";

const needsToBeReordered = [
  {
    id: "1",
    object: {
      name: "California Ale"
    },
    unitCost: 16,
    costUnit: "USD",
    currentQuantity: 1,
    quantityUnit: null,
    reorderQuantity: 2
  },
  {
    id: "2",
    object: {
      name: "English Ale"
    },
    unitCost: 10,
    costUnit: "USD",
    currentQuantity: 2,
    quantityUnit: null,
    reorderQuantity: 4
  }
];

const upcomingDeliveries = [
  {
    id: "1",
    object: {
      name: "California Ale"
    },
    unitCost: 16,
    costUnit: "USD",
    currentQuantity: 1,
    quantityUnit: null,
    reorderQuantity: 2,
    deliveryDate: "10/12/19"
  },
  {
    id: "2",
    object: {
      name: "English Ale"
    },
    unitCost: 10,
    costUnit: "USD",
    currentQuantity: 2,
    quantityUnit: null,
    reorderQuantity: 4,
    deliveryDate: "10/14/19"
  }
];

describe("Dashboard", () => {
  it("should return a FlatList", () => {
    const dashboard = shallow(<Dashboard />);
    const flatList = dashboard.find("FlatList");
    expect(flatList.length).toBe(1);
  });

  it("should have Needs to be reordered and upcoming deliveries tabs", () => {
    const dashboard = shallow(<Dashboard />);
    const tabs = dashboard.find("Tabs");
    const children = tabs.props().children().props.children;
    expect(children[0].props.value).toBe("Needs to be reordered");
    expect(children[1].props.value).toBe("Upcoming deliveries");
  });
  it("should show Needs to be reordered as active", () => {
    const dashboard = shallow(<Dashboard />);
    const tabs = dashboard.find("Tabs");
    const children = tabs.props().children().props.children;
    expect(children[0].props.active).toBe(true);
    expect(children[1].props.active).toBe(false);
  });
  it("should set the active prop to true when a tab is pressed", () => {
    const dashboard = shallow(<Dashboard />);
    const tabs = dashboard.find("Tabs");
    let children = tabs.props().children().props.children;
    children[1].props.onPress();
    dashboard.update();
    children = dashboard
      .find("Tabs")
      .props()
      .children().props.children;
    expect(children[1].props.active).toBe(true);
    expect(children[0].props.active).toBe(false);
  });
  describe("when Needs to be reordered is active", () => {
    it("should return a list of Cards equal to Needs to be Reordered data length", () => {
      const dashboard = shallow(
        <Dashboard needsToBeReordered={needsToBeReordered} />
      );
      const list = dashboard.find("FlatList").dive();
      const items = list.props().data.map(list.props().renderItem);
      expect(items.length).toBe(2);
    });
    it("should return cards with data in the correct slots", () => {
      const dashboard = renderer.create(
        <Dashboard needsToBeReordered={needsToBeReordered} />
      );
      expect(dashboard.toJSON().children).toMatchSnapshot();
    });
    it("should show total reorder cost in list footer", () => {
      const dashboard = shallow(
        <Dashboard needsToBeReordered={needsToBeReordered} />
      );
      const list = dashboard.find("FlatList");
      const value = list.props().ListFooterComponent.props.children.join("");
      expect(value).toBe("Total reorder cost: $72");
    });
  });
  describe("when Upcoming deliveries is active", () => {
    it("should return a list of Cards equal to Upcoming deliveries data length", () => {
      const dashboard = shallow(
        <Dashboard upcomingDeliveries={upcomingDeliveries} />
      );
      const tabs = dashboard.find("Tabs");
      let children = tabs.props().children().props.children;
      children[1].props.onPress();
      const list = dashboard.find("FlatList").dive();
      const items = list.props().data.map(list.props().renderItem);
      expect(items.length).toBe(2);
    });
    it("should return cards with data in the correct slots", () => {
      const dashboard = shallow(
        <Dashboard upcomingDeliveries={upcomingDeliveries} />
      );
      const tabs = dashboard.find("Tabs");
      let children = tabs.props().children().props.children;
      children[1].props.onPress();
      dashboard.update;
      const list = dashboard.find("FlatList");
      const items = list.props().data.map(list.props().renderItem);
      expect(items).toMatchSnapshot();
    });
    it("should not display a list footer", () => {
      const dashboard = shallow(
        <Dashboard upcomingDeliveries={upcomingDeliveries} />
      );
      const tabs = dashboard.find("Tabs");
      let children = tabs.props().children().props.children;
      children[1].props.onPress();
      dashboard.update();
      const list = dashboard.find("FlatList");
      const footer = list.props().ListFooterComponent;
      expect(footer).toBe(null);
    });
  });
});
