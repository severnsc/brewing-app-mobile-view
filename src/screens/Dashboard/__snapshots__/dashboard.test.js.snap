// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard when Needs to be reordered is active should return cards with data in the correct slots 1`] = `
Array [
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderBottomColor": "black",
          "borderBottomWidth": 2,
          "opacity": 1,
          "padding": 5,
        }
      }
    >
      <Text
        accessibilityLabel="Needs to be reordered"
        style={
          Object {
            "color": undefined,
            "fontWeight": "bold",
          }
        }
      >
        Needs to be reordered
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "opacity": 1,
          "padding": 5,
        }
      }
    >
      <Text
        accessibilityLabel="Upcoming deliveries"
        style={
          Object {
            "color": "#adb5bd",
            "fontWeight": "bold",
          }
        }
      >
        Upcoming deliveries
      </Text>
    </View>
  </View>,
  <FlatList
    ListFooterComponent={
      <Text>
        Total reorder cost: 
        $72
      </Text>
    }
    data={
      Array [
        Object {
          "costUnit": "USD",
          "currentQuantity": 1,
          "id": "1",
          "object": Object {
            "name": "California Ale",
          },
          "quantityUnit": null,
          "reorderQuantity": 2,
          "unitCost": 16,
        },
        Object {
          "costUnit": "USD",
          "currentQuantity": 2,
          "id": "2",
          "object": Object {
            "name": "English Ale",
          },
          "quantityUnit": null,
          "reorderQuantity": 4,
          "unitCost": 10,
        },
      ]
    }
    keyExtractor={[Function]}
    renderItem={[Function]}
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "height": 80,
          "justifyContent": "space-between",
          "opacity": 1,
          "padding": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 6,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 2,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          California Ale
        </View>
        <View />
        <View>
          $16
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          Remaining: 1
        </View>
        <View />
        <View />
        <View>
          Reorder amount: 2
        </View>
      </View>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "height": 80,
          "justifyContent": "space-between",
          "opacity": 1,
          "padding": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 6,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 2,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          English Ale
        </View>
        <View />
        <View>
          $10
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View>
          Remaining: 2
        </View>
        <View />
        <View />
        <View>
          Reorder amount: 4
        </View>
      </View>
    </View>
  </FlatList>,
]
`;

exports[`Dashboard when Upcoming deliveries is active should return cards with data in the correct slots 1`] = `
Array [
  <Card
    lowerLeft="Amount ordered: 2"
    lowerRight="Remaining: 1"
    upperLeft="California Ale"
    upperRight="10/12/19"
  />,
  <Card
    lowerLeft="Amount ordered: 4"
    lowerRight="Remaining: 2"
    upperLeft="English Ale"
    upperRight="10/14/19"
  />,
]
`;
